Realm.Shared/native/SortDescriptorBuilder.cs

```diff
@@ -0,0 +1,87 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace Realms.native
+{
+    internal class SortDescriptorBuilder
+    {
+        public class Clause
+        {
+            public List<IntPtr> ColumnIndexChain;
+            public bool Ascending;
+
+            [StructLayout(LayoutKind.Sequential)]
+            public struct Marshalable
+            {
+                public IntPtr Offset;
+                public IntPtr Count;
+
+                [MarshalAs(UnmanagedType.I1)]
+                public bool Ascending;
+            }
+        }
+
+        private readonly List<Clause> _clauses = new List<Clause>();
+        private readonly RealmObject.Metadata _metadata;
+
+        public TableHandle TableHandle => _metadata.Table;
+
+        public SortDescriptorBuilder(RealmObject.Metadata metadata)
+        {
+            _metadata = metadata;
+        }
+
+        // This method adds a direct sort clause, i.e. a clause that doesn't involve
+        // any links.
+        public void AddClause(string columnName, bool @ascending)
+        {
+            var columnIndex = _metadata.ColumnIndices[columnName];
+            _clauses.Add(new Clause {ColumnIndexChain = new List<IntPtr> { columnIndex }, Ascending = @ascending});
+        }
+
+        /// <summary>
+        /// Create a flattened array of all the clauses by concatenating the lists.
+        /// </summary>
+        /// <returns>
+        /// A tuple with two elements. Item1 is the concatenated list of indices. Item2 is the list of clauses,
+        /// indicating the offset and count, as well as the ascending flag for each clause.
+        /// </returns>
```

Comment:
This comment explains what the code does without really explaining **why** you want to flatten things. It needs to say something like:

_When you have more than one indice, all but the last one are link fields, so following the chain to sort by_ `dog.owner.name` _would need two link field indices then the index of_ `name` _in_ `Person`.